opencog/atomspace/AtomTable.cc

```diff
@@ -498,23 +495,45 @@ Handle AtomTable::add(AtomPtr atom, bool async)
     atom->keep_incoming_set();
     atom->setAtomSpace(_as);
 
+    ContentHash ch = atom->get_hash();
+
+    // Lock, to prevent two different threads from trying to add exactly
+    // the same atom.
+    std::unique_lock<std::recursive_mutex> lck(_store_mtx);
```

Comment:
And here's the bug: doing this set: `atom->setAtomSpace(_as);` is advertising that the "atom is now in the atomspace".  The test `Handle hcheck(getHandle(orig));    if (hcheck) return hcheck;` is testing this same location.  This was intended to be an atomic test-n-set, but now that it isn't, callers can get back two different copies of the same atom, both copies claiming to be in the atomspace.  Since they are two different copies, updating truth-values, counts will get scrambled, as the TV's on these two atoms will diverge over time.  All sorts of strange things from there: when its time to save, the one that is not in the atomspace won't get saved.